web/src/compose_reply.js

```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
+    if (range.commonAncestorContainer.nodeType === Node.ELEMENT_NODE) {
+        const container = range.commonAncestorContainer.cloneNode();
+        container.append(range.cloneContents());
+        outer_container.append(container);
+    } else {
+        outer_container.append(range.cloneContents());
+    }
+    return temp_div.innerHTML;
+}
+
+function get_message_selection(selection) {
+    let selected_message_content_raw = "";
+    let selection_bounds = null;
+
+    // We iterate over all ranges in the selection, to find the ranges containing
+    // the message_content div or its descendants, if any, then convert the html
+    // in those ranges to markdown for quoting (firefox can have multiple ranges
+    // in one selection), and also compute their combined bounding rect.
+    for (let i = 0; i < selection.rangeCount; i = i + 1) {
+        const range = selection.getRangeAt(i);
+        const range_common_ancestor = range.commonAncestorContainer;
+
+        // If the common ancestor is the message_content div or its child, we can quote
+        // this entire range at least.
+        if (range_common_ancestor.classList?.contains("message_content")) {
+            selected_message_content_raw =
+                selected_message_content_raw +
+                "\n" +
+                copy_and_paste.paste_handler_converter(extract_range_html(range));
+            selection_bounds = combine_bounding_rects(
+                selection_bounds,
+                range.getBoundingClientRect(),
+            );
+        } else if ($(range_common_ancestor).parents(".message_content").length) {
+            // We want to preserve the structure of the html upto 2 levels of
+            // ancestors (to retain code block / list formatting) in such a range.
+            selected_message_content_raw =
+                selected_message_content_raw +
+                "\n" +
+                copy_and_paste.paste_handler_converter(extract_range_html(range, true));
+            selection_bounds = combine_bounding_rects(
+                selection_bounds,
+                range.getBoundingClientRect(),
+            );
+        } else if (
+            // If the common ancestor contains the message_content div, we can quote the part
+            // of this range that is in the message_content div, if any.
+            range_common_ancestor instanceof Element &&
+            range_common_ancestor.querySelector(".message_content") &&
+            range.cloneContents().querySelector(".message_content")
+        ) {
+            selected_message_content_raw =
+                selected_message_content_raw +
+                "\n" +
+                copy_and_paste.paste_handler_converter(
+                    range.cloneContents().querySelector(".message_content").innerHTML,
+                );
+            selection_bounds = combine_bounding_rects(
+                selection_bounds,
+                range.getBoundingClientRect(),
+            );
+        }
+    }
```

Comment:
The three cases here all feel very similar, in a way that suggests they could be deduplicated with the use of an intermediate variable for `html_to_convert`, maybe including an `else {continue;}` block for cases where none of these `if` statements were true.

Reply:
Thanks for noticing this; initially all branches were very different, so I hadn't realised the duplication. :sweat_smile: